matrix_sdk/src/room/joined.rs

```diff
@@ -0,0 +1,455 @@
+            .map(|m| m.clone())
+        {
+            // If a group session share request is already going on,
+            // await the release of the lock.
+            mutex.lock().await;
+        } else {
+            // Otherwise create a new lock and share the group
+            // session.
+            let mutex = Arc::new(Mutex::new(()));
+            self.client
+                .group_session_locks
+                .insert(self.inner.room_id().clone(), mutex.clone());
+
+            let _guard = mutex.lock().await;
+
+            {
+                let joined = self
+                    .client
+                    .store()
+                    .get_joined_user_ids(self.inner.room_id())
+                    .await?;
+                let invited = self
+                    .client
+                    .store()
+                    .get_invited_user_ids(self.inner.room_id())
+                    .await?;
+                let members = joined.iter().chain(&invited);
+                self.client.claim_one_time_keys(members).await?;
+            };
+
+            let response = self.share_group_session().await;
+
+            self.client.group_session_locks.remove(self.inner.room_id());
+
+            // If one of the responses failed invalidate the group
+            // session as using it would end up in undecryptable
+            // messages.
+            if let Err(r) = response {
+                self.client
+                    .base_client
+                    .invalidate_group_session(self.inner.room_id())
+                    .await;
+                return Err(r);
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Share a group session for a room.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the client isn't logged in.
+    #[cfg(feature = "encryption")]
+    #[cfg_attr(feature = "docs", doc(cfg(encryption)))]
+    #[instrument]
+    async fn share_group_session(&self) -> Result<()> {
+        let mut requests = self
+            .client
+            .base_client
+            .share_group_session(self.inner.room_id())
+            .await?;
+
+        for request in requests.drain(..) {
+            let response = self.client.send_to_device(&request).await?;
+
+            self.client
+                .base_client
+                .mark_request_as_sent(&request.txn_id, &response)
+                .await?;
+        }
+
+        Ok(())
+    }
+
+    /// Send a room message to this room.
+    ///
+    /// Returns the parsed response from the server.
+    ///
+    /// If the encryption feature is enabled this method will transparently
+    /// encrypt the room message if this room is encrypted.
+    ///
+    /// # Arguments
+    ///
+    /// * `content` - The content of the message event.
+    ///
+    /// * `txn_id` - A unique `Uuid` that can be attached to a `MessageEvent`
+    /// held in its unsigned field as `transaction_id`. If not given one is
+    /// created for the message.
```

Comment:
I just realized this, why is the example gone from this method? Please, under no circumstance should the docs become worse. Ideally every method should have an example, if a method is simple enough it doesn't need to have one though I much prefer it to have one.

But *never* remove examples.

Reply:
gonna add it back